crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space/bootstrap_task.rs

```diff
@@ -0,0 +1,708 @@
+use crate::event::{KitsuneP2pEvent, KitsuneP2pEventSender, PutAgentInfoSignedEvt};
+use crate::spawn::actor::bootstrap::BootstrapNet;
+use crate::spawn::actor::space::{SpaceInternal, SpaceInternalSender};
+use crate::{KitsuneP2pResult, KitsuneSpace};
+use futures::channel::mpsc::Sender;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use ghost_actor::{GhostControlSender, GhostSender};
+use kitsune_p2p_types::agent_info::AgentInfoSigned;
+use kitsune_p2p_types::bootstrap::RandomQuery;
+use parking_lot::RwLock;
+use std::sync::Arc;
+use std::time::Duration;
+use url2::Url2;
+
+pub(super) struct BootstrapTask {
+    is_finished: bool,
+    current_delay: Duration,
+    max_delay: Duration,
+}
+
+// Trait for the bootstrap query to allow mocking in tests
+trait BootstrapService: Send {
+    fn random(&self, query: RandomQuery) -> BoxFuture<KitsuneP2pResult<Vec<AgentInfoSigned>>>;
+}
+
+struct DefaultBootstrapService {
+    url: Option<Url2>,
+    net: BootstrapNet,
+}
+
+impl BootstrapService for DefaultBootstrapService {
+    fn random(&self, query: RandomQuery) -> BoxFuture<KitsuneP2pResult<Vec<AgentInfoSigned>>> {
+        super::bootstrap::random(self.url.clone(), query, self.net).boxed()
+    }
+}
+
+impl BootstrapTask {
+    pub(super) fn spawn(
+        internal_sender: GhostSender<SpaceInternal>,
+        host_sender: Sender<KitsuneP2pEvent>,
+        space: Arc<KitsuneSpace>,
+        bootstrap_service: Option<Url2>,
+        bootstrap_net: BootstrapNet,
+        bootstrap_check_delay_backoff_multiplier: u32,
+    ) -> Arc<RwLock<Self>> {
+        let this = Arc::new(RwLock::new(BootstrapTask {
+            is_finished: false,
+            current_delay: Duration::from_secs(1),
+            max_delay: Duration::from_secs(60 * 60),
+        }));
+        let bootstrap_query = DefaultBootstrapService {
+            url: bootstrap_service,
+            net: bootstrap_net,
+        };
+
+        let task_this = this.clone();
+        BootstrapTask::spawn_inner(
+            task_this,
+            internal_sender,
+            host_sender,
+            space,
+            Box::new(bootstrap_query),
+            bootstrap_check_delay_backoff_multiplier,
+        )
+    }
+
+    fn spawn_inner(
+        this: Arc<RwLock<Self>>,
+        internal_sender: GhostSender<SpaceInternal>,
+        host_sender: Sender<KitsuneP2pEvent>,
+        space: Arc<KitsuneSpace>,
+        bootstrap_query: Box<impl BootstrapService + Send + Sync + 'static>,
+        bootstrap_check_delay_backoff_multiplier: u32,
+    ) -> Arc<RwLock<Self>> {
+        let task_this = this.clone();
+        tokio::spawn(async move {
+            let backoff_multiplier = if bootstrap_check_delay_backoff_multiplier < 2 {
```

Comment:
This is new, prevents disabling the delay by setting `bootstrap_check_delay_backoff_multiplier = 0` or disabling backoff by setting `bootstrap_check_delay_backoff_multiplier = 1`.